File: internal/maintenance/cat_junk.go
package maintenance

import (
	"context"
	"io/fs"
	"path/filepath"
	"strings"
)

func init() {
	registerMatcher("junk-dsstore", 0, scanJunkDSStore)
	registerMatcher("junk-tmp", 1, scanJunkTmp)
	registerMatcher("junk-emptydirs", 0, scanJunkEmptyDirs)
}

// junkProtectedTopLevel are root-level dirs the CLI recreates and expects to
// find; the empty-dir matcher never recurses into or offers them, and treats
// them as real (non-collapsible) content for their parent — deleting them
// buys nothing and risks a CLI mkdir race.
var junkProtectedTopLevel = map[string]bool{"projects": true, "todos": true, "plugins": true}

// walkRootBounded walks spec.Root depth-bounded (mirrors scan.go's
// maxScanDepth), Lstat-only, and skips spec.AppData's subtree entirely — the
// app's own trash/manifests must never be swept as user junk. visit runs for
// every non-root entry (files and dirs); a symlinked entry is never passed to
// visit, matching this package's symlink-refusal invariant.
func walkRootBounded(ctx context.Context, spec CategorySpec, visit func(path string, d fs.DirEntry) error) error {
	root := spec.Root
	return filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if ctxErr := ctx.Err(); ctxErr != nil {
			return ctxErr
		}
		if err != nil {
			return nil // skip unreadable entries, keep walking the rest
		}
		if path == root {
			return nil
		}
		if spec.AppData != "" && filepath.Clean(path) == filepath.Clean(spec.AppData) {
			return fs.SkipDir // never sweep the app's own data dir
		}
		if d.Type()&fs.ModeSymlink != 0 {
			return nil // never traverse; WalkDir won't recurse into it either way
		}
		if d.IsDir() {
			rel, relErr := filepath.Rel(root, path)
			if relErr == nil && strings.Count(rel, string(filepath.Separator)) > maxScanDepth {
				return fs.SkipDir
			}
		}
		return visit(path, d)
	})
}

// scanJunkDSStore flags every macOS .DS_Store file anywhere under the
// effective root. No age gate — a .DS_Store is worthless the moment it's
// written.
func scanJunkDSStore(ctx context.Context, spec CategorySpec) ([]Candidate, error) {
	out := []Candidate{}
	err := walkRootBounded(ctx, spec, func(path string, d fs.DirEntry) error {
		if d.IsDir() || d.Name() != ".DS_Store" {
			return nil
		}
		info, infoErr := d.Info()
		if infoErr != nil {
			return nil
		}
		out = append(out, Candidate{
			Path: path, Bytes: info.Size(), Files: 1, ModTime: info.ModTime(),
			Reason: "macOS metadata file",
		})
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

// scanJunkTmp flags *.tmp files older than the cutoff (default 1 day) — the
// age floor guards against sweeping a temp+rename write still in progress.
func scanJunkTmp(ctx context.Context, spec CategorySpec) ([]Candidate, error) {
	out := []Candidate{}
	err := walkRootBounded(ctx, spec, func(path string, d fs.DirEntry) error {
		if d.IsDir() || filepath.Ext(d.Name()) != ".tmp" {
			return nil
		}
		info, infoErr := d.Info()
		if infoErr != nil {
			return nil
		}
		if !olderThan(info.ModTime(), spec) {
			return nil
		}
		out = append(out, Candidate{
			Path: path, Bytes: info.Size(), Files: 1, ModTime: info.ModTime(),
			Reason: "stale temp file",
		})
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

// scanJunkEmptyDirs reports the TOPMOST fully-empty ancestor of every empty
// subtree as ONE candidate — required so a batch never asks the trash engine
// to move both a dir and something nested inside it. "Empty" means: contains
// nothing, or only a .DS_Store file and/or recursively-empty subdirs.
func scanJunkEmptyDirs(ctx context.Context, spec CategorySpec) ([]Candidate, error) {
	out := []Candidate{}
	if _, err := emptyDirScan(ctx, spec, spec.Root, 0, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// emptyDirScan is a bottom-up (post-order) walk that can't be expressed as a
// simple top-down visit callback like walkRootBounded: whether a dir gets
// reported depends on what its subdirs decide, which is only known after
// recursing into them. It mirrors the same invariants (Lstat-only/no symlink
// traversal, AppData excluded, depth-bounded, ctx.Err() checked) by hand.
//
// Returns whether dir itself is empty. When dir is NOT empty (real content,
// depth limit, or protected), any of its direct subdir children that WERE
// empty are finalized as candidates here — dir is the collapse boundary for
// that branch. When dir IS empty, nothing is finalized yet; the decision is
// deferred to whichever ancestor turns out non-empty (or the root, which is
// always non-collapsible and never itself reported).
func emptyDirScan(ctx context.Context, spec CategorySpec, dir string, depth int, out *[]Candidate) (bool, error) {
	entries, ok, err := openDirNoSymlink(dir)
	if err != nil {
		return false, err
	}
	if !ok {
		return false, nil // missing/symlink/non-dir — never empty, never a candidate
	}

	empty := true
	pendingEmpty := []string{}
	for _, e := range entries {
		if err := ctx.Err(); err != nil {
			return false, err
		}
		childPath := filepath.Join(dir, e.Name())

		if e.Type()&fs.ModeSymlink != 0 {
			empty = false // never traverse/collapse through a symlink
			continue
		}
		if !e.IsDir() {
			if e.Name() != ".DS_Store" {
				empty = false
			}
			continue
		}
		if isProtectedJunkDir(spec, dir, childPath) || depth >= maxScanDepth {
			empty = false // opaque/depth-capped: treat as real, non-collapsible content
			continue
		}

		childEmpty, childErr := emptyDirScan(ctx, spec, childPath, depth+1, out)
		if childErr != nil {
			return false, childErr
		}
		if childEmpty {
			pendingEmpty = append(pendingEmpty, childPath)
		} else {
			empty = false
		}
	}

	if empty {
		return true, nil // let the parent decide whether to collapse further
	}
	for _, p := range pendingEmpty {
		bytes, files, newest, statErr := subtreeStats(ctx, p)
		if statErr != nil {
			return false, statErr
		}
		*out = append(*out, Candidate{
			Path: p, Bytes: bytes, Files: files, ModTime: newest,
			Reason: "empty directory",
		})
	}
	return false, nil
}

// isProtectedJunkDir reports whether childPath must never be recursed into,
// collapsed, or offered: the app's own data dir, or one of the root-level
// dirs the CLI recreates and expects.
func isProtectedJunkDir(spec CategorySpec, parent, childPath string) bool {
	if spec.AppData != "" && filepath.Clean(childPath) == filepath.Clean(spec.AppData) {
		return true
	}
	if parent == spec.Root && junkProtectedTopLevel[filepath.Base(childPath)] {
		return true
	}
	return false
}

File: src-tauri/src/maintenance/cat_junk.rs
//! Ported from `internal/maintenance/cat_junk.go` (W13). `junk-dsstore`,
//! `junk-tmp`, `junk-emptydirs` + the shared depth-bounded, Lstat-only,
//! app-data-excluded root walk.

use std::collections::BTreeMap;
use std::fs::Metadata;
use std::path::{Path, PathBuf};

use super::category::{
    mtime_utc, older_than, open_dir_no_symlink, read_dir_sorted, same_path, subtree_stats,
};
use super::scan::MAX_SCAN_DEPTH;
use super::types::{Candidate, CategorySpec};

/// Root-level dirs the CLI recreates and expects; the empty-dir matcher never
/// recurses into or offers them, and treats them as real (non-collapsible)
/// content. Mirrors Go `junkProtectedTopLevel`.
fn is_protected_top_level(name: &str) -> bool {
    matches!(name, "projects" | "todos" | "plugins")
}

/// `filepath.Ext` — the suffix from the final `.` (or `""`). Note this differs
/// from `Path::extension`, which returns `None` for a dotfile like `.tmp`.
fn go_ext(name: &str) -> &str {
    match name.rfind('.') {
        Some(i) => &name[i..],
        None => "",
    }
}

fn base_name(path: &Path) -> &str {
    path.file_name().and_then(|n| n.to_str()).unwrap_or("")
}

/// Walks `spec.root` depth-bounded, Lstat-only, skipping the app-data subtree
/// entirely; `visit` runs for every non-root, non-symlink entry (files and
/// dirs). Mirrors Go `walkRootBounded`.
fn walk_root_bounded<F>(spec: &CategorySpec, mut visit: F) -> Result<(), String>
where
    F: FnMut(&Path, &Metadata) -> Result<(), String>,
{
    let root = Path::new(&spec.root);
    walk_bounded_dir(root, root, spec, &mut visit)
}

fn walk_bounded_dir<F>(
    dir: &Path,
    root: &Path,
    spec: &CategorySpec,
    visit: &mut F,
) -> Result<(), String>
where
    F: FnMut(&Path, &Metadata) -> Result<(), String>,
{
    for entry in read_dir_sorted(dir) {
        let path = entry.path();
        if !spec.app_data.is_empty() && same_path(&path, Path::new(&spec.app_data)) {
            continue; // never sweep the app's own data dir (SkipDir)
        }
        let Ok(file_type) = entry.file_type() else {
            continue;
        };
        if file_type.is_symlink() {
            continue; // never traverse
        }
        let Ok(meta) = entry.metadata() else {
            continue;
        };
        if meta.is_dir() {
            if rel_depth(root, &path) > MAX_SCAN_DEPTH {
                continue; // depth cap (SkipDir)
            }
            visit(&path, &meta)?;
            walk_bounded_dir(&path, root, spec, visit)?;
        } else {
            visit(&path, &meta)?;
        }
    }
    Ok(())
}

/// Separator count of `path` relative to `root` (Go's `strings.Count(rel, sep)`).
fn rel_depth(root: &Path, path: &Path) -> usize {
    match path.strip_prefix(root) {
        Ok(rel) => rel.components().count().saturating_sub(1),
        Err(_) => 0,
    }
}

/// Flags every macOS `.DS_Store` file under the effective root. No age gate.
/// Mirrors Go `scanJunkDSStore`.
pub fn scan_junk_dsstore(spec: &CategorySpec) -> Result<Vec<Candidate>, String> {
    let mut out: Vec<Candidate> = Vec::new();
    walk_root_bounded(spec, |path, meta| {
        if meta.is_dir() || base_name(path) != ".DS_Store" {
            return Ok(());
        }
        out.push(Candidate {
            path: path.to_string_lossy().into_owned(),
            bytes: meta.len() as i64,
            files: 1,
            mod_time: mtime_utc(meta),
            reason: "macOS metadata file".to_string(),
            group: String::new(),
            meta: BTreeMap::new(),
        });
        Ok(())
    })?;
    Ok(out)
}

/// Flags `*.tmp` files older than the cutoff (default 1 day). Mirrors Go
/// `scanJunkTmp`.
pub fn scan_junk_tmp(spec: &CategorySpec) -> Result<Vec<Candidate>, String> {
    let mut out: Vec<Candidate> = Vec::new();
    walk_root_bounded(spec, |path, meta| {
        if meta.is_dir() || go_ext(base_name(path)) != ".tmp" {
            return Ok(());
        }
        let mtime = mtime_utc(meta);
        if !older_than(mtime, spec) {
            return Ok(());
        }
        out.push(Candidate {
            path: path.to_string_lossy().into_owned(),
            bytes: meta.len() as i64,
            files: 1,
            mod_time: mtime,
            reason: "stale temp file".to_string(),
            group: String::new(),
            meta: BTreeMap::new(),
        });
        Ok(())
    })?;
    Ok(out)
}

/// Reports the TOPMOST fully-empty ancestor of every empty subtree as one
/// candidate. Mirrors Go `scanJunkEmptyDirs`.
pub fn scan_junk_empty_dirs(spec: &CategorySpec) -> Result<Vec<Candidate>, String> {
    let mut out: Vec<Candidate> = Vec::new();
    empty_dir_scan(spec, Path::new(&spec.root), 0, &mut out)?;
    Ok(out)
}

/// Bottom-up (post-order) empty-dir collapse. Returns whether `dir` itself is
/// empty; finalizes any empty direct-subdir children when `dir` is the collapse
/// boundary. Mirrors Go `emptyDirScan` (Lstat-only, no symlink traversal,
/// app-data excluded, depth-bounded). "Empty" = nothing, or only `.DS_Store`
/// files and/or recursively-empty subdirs.
fn empty_dir_scan(
    spec: &CategorySpec,
    dir: &Path,
    depth: usize,
    out: &mut Vec<Candidate>,
) -> Result<bool, String> {
    let (entries, ok) = open_dir_no_symlink(dir)?;
    if !ok {
        return Ok(false); // missing/symlink/non-dir — never empty, never a candidate
    }

    let mut empty = true;
    let mut pending_empty: Vec<PathBuf> = Vec::new();
    for e in &entries {
        let child_path = e.path();
        let file_type = match e.file_type() {
            Ok(t) => t,
            Err(_) => {
                empty = false;
                continue;
            }
        };
        if file_type.is_symlink() {
            empty = false; // never traverse/collapse through a symlink
            continue;
        }
        if !file_type.is_dir() {
            if base_name(&child_path) != ".DS_Store" {
                empty = false;
            }
            continue;
        }
        if is_protected_junk_dir(spec, dir, &child_path) || depth >= MAX_SCAN_DEPTH {
            empty = false; // opaque/depth-capped: real, non-collapsible content
            continue;
        }
        if empty_dir_scan(spec, &child_path, depth + 1, out)? {
            pending_empty.push(child_path);
        } else {
            empty = false;
        }
    }

    if empty {
        return Ok(true); // let the parent decide whether to collapse further
    }
    for p in pending_empty {
        let (bytes, files, newest) = subtree_stats(&p);
        out.push(Candidate {
            path: p.to_string_lossy().into_owned(),
            bytes,
            files,
            mod_time: newest,
            reason: "empty directory".to_string(),
            group: String::new(),
            meta: BTreeMap::new(),
        });
    }
    Ok(false)
}

/// Whether `child` must never be recursed into, collapsed, or offered: the app's
/// own data dir, or a root-level dir the CLI recreates. Mirrors Go
/// `isProtectedJunkDir`.
fn is_protected_junk_dir(spec: &CategorySpec, parent: &Path, child: &Path) -> bool {
    if !spec.app_data.is_empty() && same_path(child, Path::new(&spec.app_data)) {
        return true;
    }
    same_path(parent, Path::new(&spec.root)) && is_protected_top_level(base_name(child))
}

#[cfg(test)]
#[path = "cat_junk_tests.rs"]
mod cat_junk_tests;
